cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/index.tsx

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import IconButton from '@material-ui/core/IconButton';
+import { withStyles } from '@material-ui/core/styles';
+import AddIcon from '@material-ui/icons/Add';
+import HierarchyTree from './HierarchyTree';
+import { IWidgetProps } from 'components/AbstractWidget';
+import uuidV4 from 'uuid/v4';
+import { objectQuery } from 'services/helpers';
+import { flatToTree, toJsonHandler, jsonToFlat, getFields, inputSchemaToFlat } from './dataHandler';
+import Heading, { HeadingTypes } from 'components/Heading';
+
+export const IconWrapper = withStyles(() => {
+  return {
+    root: {
+      borderRadius: 10,
+      padding: '5px',
+      marginTop: '5px',
+      '&:focus': {
+        outline: 'none',
+      },
+    },
+  };
+})(IconButton);
+
+interface IHierarchyWidgetProps {
+  allowedTypes?: any;
+}
+
+interface IHierarchyProps extends IWidgetProps<IHierarchyWidgetProps> {
+  value: any;
+}
+
+const HierarchyWidget: React.FC<IHierarchyProps> = ({
+  value,
+  onChange,
+  extraConfig,
+  widgetProps,
+}) => {
+  const [records, setRecords]: any = React.useState([]);
+  const [tree, setTree] = React.useState([]);
+  const inputSchema = objectQuery(extraConfig, 'inputSchema') || [];
+  const allowedTypes: string[] = widgetProps.allowedTypes || [];
+  const fieldValues = getFields(inputSchema, allowedTypes);
+  const dropdownData = inputSchemaToFlat(fieldValues);
+
+  React.useEffect(() => {
+    if (value) {
+      try {
+        const responseParsed = JSON.parse(value);
+        if (responseParsed.length !== 0) {
+          const jsonToFlatData = jsonToFlat(JSON.parse(value));
+          jsonToFlatData.map((record) => {
+            if (record.path) {
+              const el = dropdownData.find((item) => item.name === record.name);
+              record.type = el.type;
+            }
+          });
+          setTree(flatToTree(jsonToFlatData));
+          setRecords(jsonToFlatData);
+        }
+      } catch (e) {
+        alert(e);
```

Comment:
We shouldn't use alert. If this is an error that user needs to understand then we need to surface this at the widget.

Some of the possible errors that the user needs to be aware
- Users using fields that are not in the input schema
- User modified the value that is no longer a valid JSON
- If the input schema is empty or undefined